File: clojure/src/step1_read_print.clj
(ns step1-read-print
    (:require [clojure.repl]
              [readline]
              [reader]
              [printer]))

;; read
(defn READ [& [strng]]
  (let [line (if strng strng (read-line))]
    (reader/read-string strng)))

;; eval
(defn EVAL [ast env]
  ast)

;; print
(defn PRINT [exp] (pr-str exp))

;; repl
(defn rep
  [strng]
  (PRINT (EVAL (READ strng) {})))

;; repl loop
(defn repl-loop []
  (let [line (readline/readline "user> ")]
    (when line
      (when-not (re-seq #"^\s*$|^\s*;.*$" line) ; blank/comment
        (try
          (println (rep line))
          (catch Throwable e
            (clojure.repl/pst e))))
      (recur))))

(defn -main [& args]
  (repl-loop))

File: fsharp/step1_read_print.fs
module REPL
    open System

    let READ input =
        try
            Reader.read_str input
        with
        | Error.ReaderError(msg) ->
            printfn "%s" msg
            []

    let EVAL ast =
        Some(ast)

    let PRINT v =
        v
        |> Seq.singleton
        |> Printer.pr_str
        |> printfn "%s"

    let REP input =
        READ input
        |> Seq.ofList
        |> Seq.map (fun form -> EVAL form)
        |> Seq.filter Option.isSome
        |> Seq.iter (fun value -> PRINT value.Value)

    let getReadlineMode args =
        if args |> Array.exists (fun e -> e = "--raw") then
            Readline.Mode.Raw
        else
            Readline.Mode.Terminal

    [<EntryPoint>]
    let rec main args =
        let mode = getReadlineMode args
        match Readline.read "user> " mode with
        | null -> 0
        | input -> 
            REP input
            main args
